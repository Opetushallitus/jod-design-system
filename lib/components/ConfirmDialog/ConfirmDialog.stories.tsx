import type { StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import type { TitledMeta } from '../../utils';

import { ConfirmDialog } from './ConfirmDialog';

const meta = {
  title: 'Popups/ConfirmDialog',
  component: ConfirmDialog,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof ConfirmDialog>;

export default meta;

type Story = StoryObj<typeof meta>;

const render = (args: Story['args']) => {
  return (
    <ConfirmDialog {...args}>
      {(showDialog) => (
        <button onClick={showDialog} className="ds:cursor-pointer">
          Open ConfirmDialog example
        </button>
      )}
    </ConfirmDialog>
  );
};

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-7677&t=b8o6NAta57e3aj8E-1',
  },
};

export const Normal: Story = {
  render,
  parameters,
  args: {
    children: () => <></>,
    title: 'Haluatko kirjautua ulos?',
    description: 'Et voi enää nähdä tiettyjä tietoja, kuten henkilökohtaisia tietoja, kun olet kirjautunut ulos.',
    onConfirm: fn(),
    confirmText: 'Kyllä',
    cancelText: 'Peruuta',
  },
};

const Content = () => (
  <div className="ds:flex ds:items-center ds:rounded-[10px] ds:border-[5px] ds:border-secondary-1 ds:px-[108px] ds:py-[22px] ds:text-center ds:text-secondary-1">
    Raahaa kuva tähän tai klikkaa aluetta valitaksesi kuva tietokoneeltasi.
  </div>
);

export const WithContent: Story = {
  render,
  parameters,
  args: {
    children: () => <></>,
    title: 'Lisää avatar-kuva',
    description: 'Lisää avatar-kuva, joka näkyy profiilissasi.',
    content: <Content />,
    onConfirm: fn(),
    confirmText: 'Lisää',
    cancelText: 'Peruuta',
  },
};

export const Destructive: Story = {
  render,
  parameters,
  args: {
    children: () => <></>,
    title: 'Haluatko poistaa suosikin?',
    description:
      'Lorem ipsum dolor sit amet, no vis verear commodo. Vix quot dicta phaedrum ad. Has eu invenire concludaturque, simul accusata no ius.',
    onConfirm: fn(),
    confirmText: 'Kyllä',
    variant: 'destructive',
    cancelText: 'Peruuta',
  },
};

const Button = ({ label, onClick }: { label: string; onClick?: () => void }) => {
  return (
    <button
      className="ds:cursor-pointer ds:group ds:flex ds:select-none ds:items-center ds:gap-4 ds:rounded-[30px] ds:bg-white ds:px-6 ds:text-button-md ds:text-black ds:hover:text-accent ds:focus-visible:text-accent ds:focus-visible:outline ds:focus-visible:outline-[3px] ds:focus-visible:outline-offset-[1.5px] ds:focus-visible:outline-accent ds:active:bg-accent ds:active:text-white ds:active:outline-0"
      onClick={onClick}
    >
      <span className="ds:py-[10px] ds:group-hover:underline ds:group-focus-visible:no-underline ds:group-active:no-underline">
        {label}
      </span>
    </button>
  );
};

const renderFooter = (args: Story['args']) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { onConfirm, confirmText, variant, cancelText, ...rest } = args;
  return (
    <ConfirmDialog
      {...rest}
      footer={(hideDialog) => (
        <div className="ds:flex ds:flex-row ds:justify-between ds:gap-5">
          <Button onClick={hideDialog} label="En" />
          <Button
            onClick={() => {
              fn();
              hideDialog();
            }}
            label="Kyllä"
          />
        </div>
      )}
    >
      {(showDialog) => (
        <button onClick={showDialog} className="ds:cursor-pointer">
          Open ConfirmDialog example
        </button>
      )}
    </ConfirmDialog>
  );
};

export const WithFooter: Story = {
  render: renderFooter,
  parameters,
  args: {
    children: () => <></>,
    footer: () => <></>,
    title: 'Haluatko kokeilla uutta ominaisuutta?',
    description:
      'Lorem ipsum dolor sit amet, no vis verear commodo. Vix quot dicta phaedrum ad. Has eu invenire concludaturque, simul accusata no ius.',
  },
};
