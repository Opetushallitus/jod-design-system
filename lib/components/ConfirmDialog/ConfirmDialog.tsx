import { Description, Dialog, DialogPanel, DialogTitle } from '@headlessui/react';
import React from 'react';
import { cx } from '../../cva';

type Variant = 'normal' | 'destructive';

type FooterProps =
  | {
      footer?: never;

      onConfirm: () => void;
      confirmText: string;
      variant?: Variant;
      cancelText: string;
    }
  | {
      footer: (closeDialog: () => void) => React.ReactNode;

      onConfirm?: never;
      confirmText?: never;
      variant?: never;
      cancelText?: never;
    };

export type ConfirmDialogProps = {
  children: (showDialog: () => void) => JSX.Element;
  title: string;
  description: string;
  content?: React.ReactNode;
} & FooterProps;

const getVariantClassNames = ({ variant }: { variant: Variant }) => {
  return cx({
    'text-alert bg-white hover:text-alert active:text-white active:bg-alert focus-visible:text-alert':
      variant === 'destructive',
    'text-black bg-white': variant === 'normal',
  });
};

const Button = ({ label, onClick, variant = 'normal' }: { label: string; onClick?: () => void; variant?: Variant }) => {
  const variantClassNames = getVariantClassNames({ variant });
  return (
    <button
      className={`group flex select-none items-center gap-4 rounded-[30px] px-6 text-button-md hover:text-accent focus-visible:text-accent focus-visible:outline focus-visible:outline-[3px] focus-visible:outline-offset-[1.5px] focus-visible:outline-accent active:bg-accent active:text-white active:outline-0 ${variantClassNames}`.trim()}
      onClick={onClick}
    >
      <span className="py-[10px] group-hover:underline group-focus-visible:no-underline group-active:no-underline">
        {label}
      </span>
    </button>
  );
};

export const ConfirmDialog = ({
  children,
  title,
  description,
  content,
  footer,
  onConfirm,
  confirmText,
  variant = 'normal',
  cancelText,
}: ConfirmDialogProps) => {
  const [isOpen, setIsOpen] = React.useState(false);

  const showDialog = () => {
    setIsOpen(true);
  };

  const hideDialog = () => {
    setIsOpen(false);
  };

  const confirmHandler = () => {
    onConfirm && onConfirm();
    hideDialog();
  };

  return (
    <>
      {children(showDialog)}
      {isOpen && (
        <Dialog open={isOpen} onClose={hideDialog} className="relative z-50">
          <div className="fixed inset-0 bg-black/30" aria-hidden />
          <div className="fixed inset-0 flex w-screen py-[96px]">
            <div className="flex w-full items-center justify-center">
              <DialogPanel className="overflow-hidden rounded-lg bg-bg-gray">
                <div className="flex max-w-[640px] flex-col p-9">
                  <DialogTitle className="mb-5 text-heading-3 font-poppins">{title}</DialogTitle>
                  <Description className="text-body-sm">{description}</Description>
                  {content && <div className="mt-7">{content}</div>}
                </div>

                <div className="overflow-x-auto overflow-y-hidden bg-bg-gray-2">
                  <div className="flex flex-row gap-5 px-6 py-5 sm:px-9 justify-end">
                    {footer ? (
                      footer(hideDialog)
                    ) : (
                      <>
                        <Button label={cancelText} onClick={hideDialog} />
                        <Button label={confirmText} onClick={confirmHandler} variant={variant} />
                      </>
                    )}
                  </div>
                </div>
              </DialogPanel>
            </div>
          </div>
        </Dialog>
      )}
    </>
  );
};
