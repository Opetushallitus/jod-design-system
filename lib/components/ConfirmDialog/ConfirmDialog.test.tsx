import { act, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { userEvent } from '@storybook/test';
import { ConfirmDialog } from './ConfirmDialog';

describe('ConfirmDialog', () => {
  it('renders dialog when showDialog is called', async () => {
    const user = userEvent.setup();
    render(
      <ConfirmDialog
        title="ConfirmDialog rendered"
        description="Are you sure?"
        onConfirm={vi.fn()}
        confirmText="Confirm"
        cancelText="Cancel"
      >
        {(showDialog) => <button onClick={showDialog}>Show for rendering</button>}
      </ConfirmDialog>,
    );

    await act(async () => {
      await user.click(screen.getByText('Show for rendering'));
    });

    expect(screen.getByText('ConfirmDialog rendered')).toBeInTheDocument();
    expect(screen.getByText('Are you sure?')).toBeInTheDocument();
    expect(screen.getByLabelText('ConfirmDialog rendered')).toMatchSnapshot();
  });

  it('calls onConfirm when confirm button is clicked', async () => {
    const onConfirmMock = vi.fn();
    const user = userEvent.setup();
    render(
      <ConfirmDialog
        title="onConfirm called"
        description="Are you sure?"
        onConfirm={onConfirmMock}
        confirmText="Confirm"
        cancelText="Cancel"
      >
        {(showDialog) => <button onClick={showDialog}>Show for onConfirm</button>}
      </ConfirmDialog>,
    );
    await act(async () => {
      await user.click(screen.getByText('Show for onConfirm'));
    });
    await act(async () => {
      await user.click(screen.getByText('Confirm'));
    });

    expect(onConfirmMock).toHaveBeenCalled();
  });

  it('closes dialog when cancel button is clicked', async () => {
    const user = userEvent.setup();
    render(
      <ConfirmDialog
        title="ConfirmDialog closes"
        description="Are you sure?"
        onConfirm={vi.fn()}
        confirmText="Confirm"
        cancelText="Cancel"
      >
        {(showDialog) => <button onClick={showDialog}>Test closing</button>}
      </ConfirmDialog>,
    );

    await act(async () => {
      await user.click(screen.getByText('Test closing'));
    });
    await act(async () => {
      await user.click(screen.getByText('Cancel'));
    });

    expect(screen.queryByText('ConfirmDialog closes')).not.toBeInTheDocument();
    expect(screen.queryByText('Are you sure?')).not.toBeInTheDocument();
  });
});
