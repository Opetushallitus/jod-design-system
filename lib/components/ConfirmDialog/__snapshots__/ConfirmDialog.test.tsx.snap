// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmDialog > renders dialog when showDialog is called 1`] = `
<div
  aria-describedby="headlessui-description-«ra»"
  aria-labelledby="headlessui-dialog-title-«r9»"
  aria-modal="true"
  class="ds:relative ds:z-50"
  data-headlessui-state="open"
  data-open=""
  id="ds-confirm-dialog-«r0»"
  role="dialog"
  tabindex="-1"
>
  <div
    aria-hidden="true"
    class="ds:fixed ds:inset-0 ds:bg-black/30"
  />
  <div
    class="ds:fixed ds:inset-0 ds:flex ds:w-screen ds:py-[96px]"
  >
    <div
      class="ds:flex ds:w-full ds:items-center ds:justify-center"
    >
      <div
        class="ds:overflow-hidden ds:rounded-lg ds:bg-bg-gray"
        data-headlessui-state="open"
        data-open=""
        id="ds-confirm-dialog-panel-«r0»"
      >
        <div
          class="ds:flex ds:max-w-[640px] ds:flex-col ds:p-9"
        >
          <h2
            class="ds:mb-5 ds:text-heading-2"
            data-headlessui-state="open"
            data-open=""
            id="headlessui-dialog-title-«r9»"
          >
            ConfirmDialog rendered
          </h2>
          <p
            class="ds:text-body-sm ds:font-arial"
            data-headlessui-state="open"
            data-open=""
            id="headlessui-description-«ra»"
          >
            Are you sure?
          </p>
        </div>
        <div
          class="ds:overflow-x-auto ds:overflow-y-hidden ds:bg-bg-gray-2"
        >
          <div
            class="ds:flex ds:flex-row ds:gap-5 ds:px-6 ds:py-5 ds:sm:px-9 ds:justify-end"
          >
            <button
              class="ds:cursor-pointer ds:group ds:flex ds:select-none ds:items-center ds:gap-4 ds:rounded-[30px] ds:px-6 ds:text-button-md ds:hover:text-accent ds:focus-visible:text-accent ds:focus-visible:outline ds:focus-visible:outline-[3px] ds:focus-visible:outline-offset-[1.5px] ds:focus-visible:outline-accent ds:active:bg-accent ds:active:text-white ds:active:outline-0 ds:text-black ds:bg-white"
            >
              <span
                class="ds:py-[10px] ds:group-hover:underline ds:group-focus-visible:no-underline ds:group-active:no-underline"
              >
                Cancel
              </span>
            </button>
            <button
              class="ds:cursor-pointer ds:group ds:flex ds:select-none ds:items-center ds:gap-4 ds:rounded-[30px] ds:px-6 ds:text-button-md ds:hover:text-accent ds:focus-visible:text-accent ds:focus-visible:outline ds:focus-visible:outline-[3px] ds:focus-visible:outline-offset-[1.5px] ds:focus-visible:outline-accent ds:active:bg-accent ds:active:text-white ds:active:outline-0 ds:text-black ds:bg-white"
            >
              <span
                class="ds:py-[10px] ds:group-hover:underline ds:group-focus-visible:no-underline ds:group-active:no-underline"
              >
                Confirm
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
